server/src/main/java/org/elasticsearch/health/GetHealthAction.java

```diff
@@ -0,0 +1,168 @@
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.startObject();
+            builder.field("status", status);
+            builder.field("cluster_name", clusterName.value());
+            builder.array("impacts");
+            builder.startObject("components");
+            for (Component component : components) {
+                builder.field(component.name());
+                component.toXContent(builder, params);
+            }
+            builder.endObject();
+            return builder.endObject();
+        }
+    }
+
+    public static class Request extends ActionRequest {
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    public static class TransportAction extends org.elasticsearch.action.support.TransportAction<Request, Response> {
+
+        private final ClusterService clusterService;
+
+        @Inject
+        public TransportAction(
+            final ActionFilters actionFilters,
+            final TransportService transportService,
+            final ClusterService clusterService
+        ) {
+            super(NAME, actionFilters, transportService.getTaskManager());
+            this.clusterService = clusterService;
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+            final ClusterState clusterState = clusterService.state();
+            final Component controller = Controller.createControllerComponent(clusterService.localNode(), clusterState);
+            final Component snapshots = new Component("snapshots", HealthStatus.GREEN, Collections.emptyMap());
+            final ClusterName clusterName = clusterService.getClusterName();
+            listener.onResponse(new Response(clusterName, Arrays.asList(controller, snapshots)));
+        }
+    }
+
+    public record Component(String name, HealthStatus status, Map<String, Indicator> indicators) implements ToXContentObject {
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("status", status);
+            builder.startObject("indicators");
+            for (Map.Entry<String, Indicator> indicator : indicators.entrySet()) {
+                builder.field(indicator.getKey());
+                indicator.getValue().toXContent(builder, params);
+            }
+            builder.endObject();
+            return builder.endObject();
+        }
+
+    }
+
+    public record Indicator(String name, HealthStatus status, String summary, ToXContentFragment details) implements ToXContentObject {
```

Comment:
Not sure if this was discussed already, but is it worth making Indicator (and may be Component) a top level classes?
Looks like this could be return type of the HealthIndicatorService from the plugins and it would be surprising to import something that is nested to an action